sql/importer: TestImportIntoCSVCancel failed

sql/importer.TestImportIntoCSVCancel [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16216844?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16216844?buildTab=artifacts#/) on release-24.2 @ [01e09cbd72ec2e2e35ea0ba190eb4931fc1e19e7](https://github.com/cockroachdb/cockroach/commits/01e09cbd72ec2e2e35ea0ba190eb4931fc1e19e7):


```
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x400c2f6e98?, {0x7b5ba50, 0x400dd40120})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x68
panic({0x57a2fc0?, 0xb5f2e00?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/sql/importer.(*readImportDataProcessor).close(0x4009233b08)
	github.com/cockroachdb/cockroach/pkg/sql/importer/import_processor.go:241 +0x34
github.com/cockroachdb/cockroach/pkg/sql/importer.(*readImportDataProcessor).ConsumerClosed(0x40074cb868?)
	github.com/cockroachdb/cockroach/pkg/sql/importer/import_processor.go:232 +0x1c
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Close(...)
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:745
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0x40074cb688, {0x7b5ba18, 0x4007ecc550})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:682 +0x70
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0x400dd40840, {0x7b5ba18, 0x4007ecc550})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0x10c
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow.func1(0x4007819008, 0x400c2f73e0, 0x4008d30808, 0x400940aa80, {0x7b5ba50, 0x400dd40120})
	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:653 +0x15c
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow(0x4007819008, {0x7b5ba18, 0x4007ecc550}, 0x400940aa80)
	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:667 +0x158
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1({0x7b5ba50?, 0x400dd40120?}, {0x5f16300?, 0x400940aa80?})
	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/aarch64-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:597 +0xd0
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.ServerInterceptor.func12({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80}, 0x401066e960, 0x4007677680)
	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97 +0x32c
google.golang.org/grpc.getChainUnaryHandler.func1({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xa0
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80}, 0x401066e960?, 0x40020b46c0)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169 +0x88
google.golang.org/grpc.getChainUnaryHandler.func1({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xa0
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor({0x4007640000?, {{0x5d8a2c0?}, {0x7baa418?, 0x4007c854a0?}}}, {0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80}, 0x401066e960, 0x40020b4600)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:95 +0x15c
google.golang.org/grpc.getChainUnaryHandler.func1({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xa0
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1({0x7b5ba50?, 0x400dd40120?})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:136 +0x40
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0x4004b4efc0, {0x7b5ba50, 0x400dd40120}, {0xba2f5c?, 0x38?}, 0x4010753960)
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:337 +0xac
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1({0x7b5ba50?, 0x400dd40120?}, {0x5f16300?, 0x400940aa80?}, 0x401066e960?, 0x4007677680?)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:134 +0x7c
google.golang.org/grpc.NewServer.chainUnaryServerInterceptors.chainUnaryInterceptors.func1({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80}, 0x401066e960, 0x5ee0de0?)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1145 +0x88
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler({0x5ff7e80, 0x4007819008}, {0x7b5ba50, 0x400dd40120}, 0x400315e690, 0x4008398c80)
	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/aarch64-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:599 +0x148
google.golang.org/grpc.(*Server).processUnaryRPC(0x40098045a0, {0x7bb5000, 0x400b156820}, 0x4008918900, 0x4001e7e060, 0xb6a2400, 0x0)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1335 +0xb90
google.golang.org/grpc.(*Server).handleStream(0x40098045a0, {0x7bb5000, 0x400b156820}, 0x4008918900, 0x0)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1712 +0x7ac
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:947 +0xb4
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 7082
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:958 +0x14c
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestImportIntoCSVCancel.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-40566

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
      [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x42a49e4]
      
      goroutine 17515 [running]:
      github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x400c2f6e98?, {0x7b5ba50, 0x400dd40120})
        github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x68
      panic({0x57a2fc0?, 0xb5f2e00?})
        GOROOT/src/runtime/panic.go:770 +0x124
      github.com/cockroachdb/cockroach/pkg/sql/importer.(*readImportDataProcessor).close(0x4009233b08)
        github.com/cockroachdb/cockroach/pkg/sql/importer/import_processor.go:241 +0x34
      github.com/cockroachdb/cockroach/pkg/sql/importer.(*readImportDataProcessor).ConsumerClosed(0x40074cb868?)
        github.com/cockroachdb/cockroach/pkg/sql/importer/import_processor.go:232 +0x1c
      github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Close(...)
        github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:745
      github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0x40074cb688, {0x7b5ba18, 0x4007ecc550})
        github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:682 +0x70
      github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0x400dd40840, {0x7b5ba18, 0x4007ecc550})
        github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0x10c
      github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow.func1(0x4007819008, 0x400c2f73e0, 0x4008d30808, 0x400940aa80, {0x7b5ba50, 0x400dd40120})
        github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:653 +0x15c
      github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow(0x4007819008, {0x7b5ba18, 0x4007ecc550}, 0x400940aa80)
        github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:667 +0x158
      github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1({0x7b5ba50?, 0x400dd40120?}, {0x5f16300?, 0x400940aa80?})
        github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/aarch64-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:597 +0xd0
      github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.ServerInterceptor.func12({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80}, 0x401066e960, 0x4007677680)
        github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97 +0x32c
      google.golang.org/grpc.getChainUnaryHandler.func1({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80})
        google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xa0
      github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80}, 0x401066e960?, 0x40020b46c0)
        github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169 +0x88
      google.golang.org/grpc.getChainUnaryHandler.func1({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80})
        google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xa0
      github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor({0x4007640000?, {{0x5d8a2c0?}, {0x7baa418?, 0x4007c854a0?}}}, {0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80}, 0x401066e960, 0x40020b4600)
        github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:95 +0x15c
      google.golang.org/grpc.getChainUnaryHandler.func1({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80})
        google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xa0
      github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1({0x7b5ba50?, 0x400dd40120?})
        github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:136 +0x40
      github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0x4004b4efc0, {0x7b5ba50, 0x400dd40120}, {0xba2f5c?, 0x38?}, 0x4010753960)
        github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:337 +0xac
      github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1({0x7b5ba50?, 0x400dd40120?}, {0x5f16300?, 0x400940aa80?}, 0x401066e960?, 0x4007677680?)
        github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:134 +0x7c
      google.golang.org/grpc.NewServer.chainUnaryServerInterceptors.chainUnaryInterceptors.func1({0x7b5ba50, 0x400dd40120}, {0x5f16300, 0x400940aa80}, 0x401066e960, 0x5ee0de0?)
        google.golang.org/grpc/external/org_golang_google_grpc/server.go:1145 +0x88
      github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler({0x5ff7e80, 0x4007819008}, {0x7b5ba50, 0x400dd40120}, 0x400315e690, 0x4008398c80)
        github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/aarch64-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:599 +0x148
      google.golang.org/grpc.(*Server).processUnaryRPC(0x40098045a0, {0x7bb5000, 0x400b156820}, 0x4008918900, 0x4001e7e060, 0xb6a2400, 0x0)
        google.golang.org/grpc/external/org_golang_google_grpc/server.go:1335 +0xb90
      google.golang.org/grpc.(*Server).handleStream(0x40098045a0, {0x7bb5000, 0x400b156820}, 0x4008918900, 0x0)
        google.golang.org/grpc/external/org_golang_google_grpc/server.go:1712 +0x7ac
      google.golang.org/grpc.(*Server).serveStreams.func1.1()
        google.golang.org/grpc/external/org_golang_google_grpc/server.go:947 +0xb4
      created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 7082
        google.golang.org/grpc/external/org_golang_google_grpc/server.go:958 +0x14c
```

--- a/pkg/ccl/backupccl/backup_processor.go
+++ b/pkg/ccl/backupccl/backup_processor.go
@@ -288,7 +288,9 @@ func (bp *backupDataProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.Producer
 }
 
 func (bp *backupDataProcessor) close() {
-	bp.cancelAndWaitForWorker()
+	if bp.cancelAndWaitForWorker != nil {
+		bp.cancelAndWaitForWorker()
+	}
 	if bp.InternalClose() {
 		bp.aggTimer.Stop()
 		bp.memAcc.Close(bp.Ctx())

--- a/pkg/ccl/backupccl/generative_split_and_scatter_processor.go
+++ b/pkg/ccl/backupccl/generative_split_and_scatter_processor.go
@@ -393,7 +393,9 @@ func (gssp *generativeSplitAndScatterProcessor) ConsumerClosed() {
 // runs into an error and stops consuming scattered entries to make sure we
 // don't leak goroutines.
 func (gssp *generativeSplitAndScatterProcessor) close() {
-	gssp.cancelScatterAndWaitForWorker()
+	if gssp.cancelScatterAndWaitForWorker != nil {
+		gssp.cancelScatterAndWaitForWorker()
+	}
 	gssp.InternalClose()
 }
 

--- a/pkg/ccl/backupccl/restore_data_processor.go
+++ b/pkg/ccl/backupccl/restore_data_processor.go
@@ -667,7 +667,9 @@ func (rd *restoreDataProcessor) ConsumerClosed() {
 	if rd.Closed {
 		return
 	}
-	rd.cancelWorkersAndWait()
+	if rd.cancelWorkersAndWait != nil {
+		rd.cancelWorkersAndWait()
+	}
 
 	rd.qp.Close(rd.Ctx())
 	rd.aggTimer.Stop()

--- a/pkg/sql/execinfra/base.go
+++ b/pkg/sql/execinfra/base.go
@@ -123,7 +123,7 @@ type RowSource interface {
 
 	// Start prepares the RowSource for future Next() calls and takes in the
 	// context in which these future calls should operate. Start needs to be
-	// called before Next/ConsumerDone/ConsumerClosed.
+	// called before Next and ConsumerDone.
 	//
 	// RowSources that consume other RowSources are expected to Start() their
 	// inputs.
@@ -169,7 +169,8 @@ type RowSource interface {
 
 	// ConsumerClosed informs the source that the consumer is done and will not
 	// make any more calls to Next(). Must be called at least once on a given
-	// RowSource and can be called multiple times.
+	// RowSource and can be called multiple times. Implementations must support
+	// the case when Start was never called.
 	//
 	// Like ConsumerDone(), if the consumer of the source stops consuming rows
 	// before Next indicates that there are no more rows, ConsumerDone() and/or

--- a/pkg/sql/importer/import_processor.go
+++ b/pkg/sql/importer/import_processor.go
@@ -238,7 +238,9 @@ func (idp *readImportDataProcessor) close() {
 		return
 	}
 
-	idp.cancel()
+	if idp.cancel != nil {
+		idp.cancel()
+	}
 	_ = idp.wg.Wait()
 
 	idp.InternalClose()

